src/seccompiler/src/types.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+use std::ffi::CString;
+use std::str::FromStr;
+
+use serde::*;
+
+// use libseccomp::{ScmpAction, ScmpArch, ScmpCompareOp};
+use crate::bindings::*;
+
+/// Comparison to perform when matching a condition.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum SeccompCmpOp {
+    Eq,
+    Ge,
+    Gt,
+    Le,
+    Lt,
+    MaskedEq(u64),
+    Ne,
+}
+
+/// Condition that syscall must match in order to satisfy a rule.
+#[derive(Debug, Deserialize)]
+pub struct SeccompCondition {
+    pub index: u8,
+    pub op: SeccompCmpOp,
+    pub val: u64,
+}
+
+impl SeccompCondition {
+    pub fn to_scmp_type(&self, syscall: i32) -> scmp_arg_cmp {
+        #[cfg(target_arch = "x86_64")]
+        const IOCTL: i32 = 16;
+        #[cfg(target_arch = "aarch64")]
+        const IOCTL: i32 = 29;
```

Comment:
done

Reply:
doesnt libseccomp have this cool function for converting between syscall name and number? or is that only available in python? :o

Reply:
yeah in C there's two: https://github.com/seccomp/libseccomp/blob/main/include/seccomp.h.in#L609-L631

Reply:
But we already have the string version, we get it from the JSON file. So I think we could just pass the syscall string instead of the number and then we don't need to do any re-translation.

Reply:
You okay if I commit this diff?

```
diff --git a/src/seccompiler/src/bindings.rs b/src/seccompiler/src/bindings.rs
index c13797ef31a4..bf6214f55473 100644
--- a/src/seccompiler/src/bindings.rs
+++ b/src/seccompiler/src/bindings.rs
@@ -163,6 +163,18 @@ unsafe extern "C" {
     /// This function generates seccomp Berkeley Packer Filter (BPF) code and writes
     /// it to the given fd.  Returns zero on success, negative values on failure.
     pub fn seccomp_export_bpf(ctx: const_scmp_filter_ctx, fd: c_int) -> c_int;
+
+    ///
+    /// Resolve a syscall name to a number
+    /// @param arch_token the architecture token, e.g. SCMP_ARCH_*
+    /// @param name the syscall name
+    ///
+    /// Resolve the given syscall name to the syscall number for the given
+    /// architecture.  Returns the syscall number on success, including negative
+    /// pseudo syscall numbers (e.g. __PNR_///); returns __NR_SCMP_ERROR on failure.
+    ///
+    ///
+    pub fn seccomp_syscall_resolve_name_arch(arch_token: u32, name: *const libc::c_char) -> c_int;
 }
 
 /// Negative pseudo syscall number returned by some functions in case of an error
diff --git a/src/seccompiler/src/types.rs b/src/seccompiler/src/types.rs
index 449b0438fa86..d0cddfa55952 100644
--- a/src/seccompiler/src/types.rs
+++ b/src/seccompiler/src/types.rs
@@ -36,10 +36,11 @@ impl SeccompCondition {
         // For `ioctls` we need to mask upper bits as musl
         // sets them to 1, but libseccomp explicitly checks that they are 0.
         // with 0x00000000FFFFFFFF mask upper bits are always 0.
-        let ioctl = match arch {
-            TargetArch::X86_64 => 16,
-            TargetArch::Aarch64 => 29,
+        // SAFETY: c"ioctl".as_ptr() is a pointer to a nul-terminated C-string.
+        let ioctl = unsafe {
+            seccomp_syscall_resolve_name_arch(arch.to_scmp_type(), c"ioctl".as_ptr())
         };
+
         match self.op {
             SeccompCmpOp::Eq => {
                 if syscall == ioctl {

```

Reply:
Ah, you're right... we first translate from string to number, only to then kinda need the string here again

Reply:
Actually, instead of hotpatching these in seccompiler-bin, maybe we should just update our seccomp filters for ioctls to use masked_eq? :sweat_smile: